In [84]:
import pandas as pd

In [85]:
import numpy as np

In [8]:
df=pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv'
)

In [9]:
df

Laptop       Status   Brand  \
0     ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...          New    Asus   
1     Alurin Go Start Intel Celeron N4020/8GB/256GB ...          New  Alurin   
2     ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...          New    Asus   
3     MSI Katana GF66 12UC-082XES Intel Core i7-1270...          New     MSI   
4     HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...          New      HP   
...                                                 ...          ...     ...   
2155  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2156  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2157  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2158  Razer Book 13 Intel Evo Core i7-1165G7/16GB/1T...  Refurbished   Razer   
2159  Razer Book FHD+ Intel Evo Core i7-1165G7/16GB/...  Refurbished   Razer   

           Model                CPU  RAM  Storage Storage type       GPU  \
0     ExpertBook      Intel Core i5    8      512          SSD       NaN   
1             Go      Intel Celeron    8      256          SSD       NaN   
2     ExpertBook      Intel Core i3    8      256          SSD       NaN   
3         Katana      Intel Core i7   16     1000          SSD  RTX 3050   
4            15S      Intel Core i5   16      512          SSD       NaN   
...          ...                ...  ...      ...          ...       ...   
2155       Blade      Intel Core i7   16     1000          SSD  RTX 3060   
2156       Blade      Intel Core i7   16     1000          SSD  RTX 3070   
2157       Blade      Intel Core i7   32     1000          SSD  RTX 3080   
2158        Book  Intel Evo Core i7   16     1000          SSD       NaN   
2159        Book  Intel Evo Core i7   16      256          SSD       NaN   

      Screen Touch  Final Price  
0       15.6    No      1009.00  
1       15.6    No       299.00  
2       15.6    No       789.00  
3       15.6    No      1199.00  
4       15.6    No       669.01  
...      ...   ...          ...  
2155    17.3    No      2699.99  
2156    17.3    No      2899.99  
2157    17.3    No      3399.99  
2158    13.4   Yes      1899.99  
2159    13.4   Yes      1699.99  

[2160 rows x 12 columns]

Q2. Records count
How many records are in the dataset?

Answer: The idea is to extract that information using pandas. 

In [5]:

df.shape


(2160, 12)

Esto nos indica que la cantidad de Filas es 2160 y la cantidad de columnas es 12

Q3. Laptop brands:

How many laptop brands are presented in the dataset?

For this we need to access the elements of said column. The command for this is df['brand']. If we chain the command .unique it will output the brand names. For accesing the lenght, we could use .nunique instead, or use len(df['brand'].unique)  

In [24]:
len(df['Brand'].unique())

27



Q4. Missing values:
How many columns in the dataset have missing values?




For this, we will use isnull chained with .any. isnull outputs a df with the same size of the original df where the cells are true if they're missing a value, or false if they have a value. Aditionally, any checks for that DF, if there's a column that meets that criteria. Finally, if we add .sum(), it will count every true (as python converts true to 1 and false to 0) so we will end with a sum of every time a column has a NAN or none value. Agrego int al principio para que me de solo el numero.

In [44]:
int(df.isnull().any().sum())

3

Q5. Maximum final price:
What's the maximum final price of Dell notebooks in the dataset?




Here the idea is to check all the rows that the brand is "Dell" and check for the highest price. We will using df[df['Brand'].str.contains('Dell',na=False)] to obtain a sub dataframe that contains Only dell notebooks. After that, we'will be accesing to the column Final Price (so we onw have  a series obtained from the column prices. After that, we will use .max and float so we can extract  the biggest value of said column.

In [50]:
float(df[df['Brand'].str.contains('Dell',na=False)]['Final Price'].max())

3936.0

Q6. Median value of Screen:
Find the median value of Screen column in the dataset.
Next, calculate the most frequent value of the same Screen column.
Use fillna method to fill the missing values in Screen column with the most frequent value from the previous step.
Now, calculate the median value of Screen once again.
Has it changed?

Hint: refer to existing mode and median functions to complete the task.




Now i want to access to statistical indicators of a given column. We need to isolate said column and do mode and median pd functions to extract information. We will use df['Screen'].mode() and df['Screen'].median()

In [56]:
df['Screen'].mode()

0    15.6
Name: Screen, dtype: float64

In [59]:
float(df['Screen'].median())

15.6

In [62]:
df['Screen'].fillna(df['Screen'].mode()).median()

np.float64(15.6)

Q7. Sum of weights:

Select all the "Innjoo" laptops from the dataset.

Select only columns RAM, Storage, Screen.

Get the underlying NumPy array. Let's call it X.

Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.

Compute the inverse of XTX.

Create an array y with values [1100, 1300, 800, 900, 1000, 1100].

Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.

What's the sum of all the elements of the result?

In [79]:
df[df['Brand'].str.contains('Innjoo',na=False)][['RAM','Storage','Screen']]

RAM  Storage  Screen
1478    8      256    15.6
1479    8      512    15.6
1480    4       64    14.1
1481    6       64    14.1
1482    6      128    14.1
1483    6      128    14.1

In [80]:
X=df[df['Brand'].str.contains('Innjoo',na=False)][['RAM','Storage','Screen']].values

In [82]:
X.T

array([[  8. ,   8. ,   4. ,   6. ,   6. ,   6. ],
       [256. , 512. ,  64. ,  64. , 128. , 128. ],
       [ 15.6,  15.6,  14.1,  14.1,  14.1,  14.1]])

In [86]:
XTX=np.dot(X.T,X)

In [87]:
XTX

array([[2.52000e+02, 8.32000e+03, 5.59800e+02],
       [8.32000e+03, 3.68640e+05, 1.73952e+04],
       [5.59800e+02, 1.73952e+04, 1.28196e+03]])

In [88]:
inv=np.linalg.inv(XTX)

In [89]:
inv

array([[ 2.78025381e-01, -1.51791334e-03, -1.00809855e-01],
       [-1.51791334e-03,  1.58286725e-05,  4.48052175e-04],
       [-1.00809855e-01,  4.48052175e-04,  3.87214888e-02]])

[[ 1.00000000e+00  5.21598882e-17 -7.04660473e-16]
 [-2.75571802e-13  1.00000000e+00 -2.84062325e-14]
 [ 6.32385368e-15 -1.63195767e-18  1.00000000e+00]]


In [93]:
y=[1100, 1300, 800, 900, 1000, 1100]

In [94]:
np.dot(inv,X.T)

array([[ 0.26298349, -0.12560233, -0.40646389,  0.14958687,  0.05244042,
         0.05244042],
       [-0.00110155,  0.00295059,  0.00125892, -0.00177691, -0.00076387,
        -0.00076387],
       [-0.08772226,  0.0269791 ,  0.17140891, -0.0302108 , -0.00153546,
        -0.00153546]])

In [95]:
np.dot(np.dot(inv,X.T),y)

array([45.58076606,  0.42783519, 45.29127938])

In [96]:
w=np.dot(np.dot(inv,X.T),y)

In [97]:
total=w.sum()

In [98]:
total

np.float64(91.2998806299555)